borgmatic/hooks/snapper.py

```diff
@@ -0,0 +1,137 @@
+import json
+import logging
+import os
+import shutil
+from copy import copy
+from functools import lru_cache
+from pathlib import Path
+
+from borgmatic.execute import execute_command_and_capture_output
+
+logger = logging.getLogger(__name__)
+
+
+def _call_snapper(*args) -> dict:
+    return json.loads(execute_command_and_capture_output(['snapper', '--jsonout', *args]))
+
+
+@lru_cache(maxsize=None)
+def _available_configs() -> dict[Path, str]:
+    '''
+    Returns available snapper configs mapping subvolume paths to snapper config names
+    '''
+    configs = _call_snapper('list-configs')['configs']
+    # using Path as a key makes it possible to ignore the trailing slash problem
+    # i.e. Path('/test') == Path('/test/')
+    return {Path(config['subvolume']): config['config'] for config in configs}
+
+
+def prepare_source_directories(hook_config, _log_prefix, source_directories):
+    '''
+    Computes latest snapper snapshot for each configured path based on source_directories
+    '''
+
+    if hook_config == {}:
+        return source_directories
+    source_directories = set(map(Path, source_directories))
+
+    if hook_config['include'] == 'all':
+        snapper_dirs = copy(source_directories)
+        include_all = True
+    else:
+        include_dirs = set(map(Path, hook_config['include']))
+        snapper_dirs = source_directories.intersection(include_dirs)
+        include_all = False
+
+    if hook_config.get('exclude'):
+        exclude_dirs = set(map(Path, hook_config['exclude']))
+        snapper_dirs -= exclude_dirs
```

Comment:
It occurs to me that if you lean on a Borg dry-run to produce the list of directories to process, you could even make it apply the includes and excludes for you so you don't have to here. You'd just have to pass them in on Borg's command-line.